pkg/apiserver/clusterinfo/fetch.go

```diff
@@ -0,0 +1,227 @@
+// Copyright 2020 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package clusterinfo
+
+import (
+	"context"
+	"encoding/json"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+
+	"github.com/pingcap/kvproto/pkg/metapb"
+
+	"github.com/pingcap/log"
+
+	"github.com/pkg/errors"
+
+	"github.com/pingcap-incubator/tidb-dashboard/pkg/utils/clusterinfo"
+)
+
+// fetcher is an interface for concurrently fetch data and store it in `info`.
+type fetcher interface {
+	// fetch fetches the data, and if any unrecoverable error exists, it will return error.
+	fetch(ctx context.Context, info *ClusterInfo, service *Service) error
+}
+
+// etcdFetcher fetches etcd, and parses the ns below:
+// * /topology/grafana
+// * /topology/alertmanager
+// * /topology/tidb
+type etcdFetcher struct{}
+
+func (f etcdFetcher) fetch(ctx context.Context, info *ClusterInfo, service *Service) error {
+	tidb, grafana, alertManager, err := clusterinfo.FetchEtcd(ctx, service.etcdCli)
+	if err != nil {
+		return err
+	}
+	info.TiDB = tidb
+	info.Grafana = grafana
+	info.AlertManager = alertManager
+	return nil
+}
+
+// PDFetcher using the http to fetch PDMember information from pd endpoint.
+type pdFetcher struct {
+}
+
+func (p pdFetcher) buildHealthMap(pdEndPoint string) (map[string]struct{}, error) {
+	// health member set
+	healthMember := map[string]struct{}{}
+
+	resp, err := http.Get(pdEndPoint + "/pd/api/v1/health")
+	if err != nil {
+		return nil, err
+	}
+	data, err := ioutil.ReadAll(resp.Body)
+
+	if err != nil {
+		return nil, err
+	}
+
+	var healths []struct {
+		MemberID json.Number `json:"member_id"`
+	}
+
+	err = json.Unmarshal(data, &healths)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, v := range healths {
+		healthMember[v.MemberID.String()] = struct{}{}
+	}
+	return healthMember, nil
+}
+
+func (p pdFetcher) fetch(ctx context.Context, info *ClusterInfo, service *Service) error {
+	resp, err := http.Get(service.config.PDEndPoint + "/pd/api/v1/members")
```

Comment:
Is it possible to use PD client or Etcd client here?

Reply:
I think that was no difference

Reply:
pd client have security options that is easier to use.